src/compiler/checker.ts

```diff
@@ -16856,6 +16856,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 const propType = getTypeOfSymbol(prop);
                 return accessExpression && getAssignmentTargetKind(accessExpression) !== AssignmentKind.Definite ?
                     getFlowTypeOfReference(accessExpression, propType) :
+                    accessNode && isIndexedAccessTypeNode(accessNode) ?
+                    mapType(propType, type => type === missingType ? undefinedType : type) :

```

Comment:
this is a good fix for the problem encountered here, and probably for most cases when referencing a single property, but I think it might have undesired effects for cases where the indexType is a union of property names, especially something like `keyof T`.

Am I wrong? Does it still make sense to turn missing into undefined there too? @weswigham I'd like your advice on this.

Reply:
But even in such cases - we just end up distributing~ the indexType over objectType. Essentially it's the same as `T['a'] | T['b'] | ...`. So the open question is - is there any case for which indexed access type should "transfer" the optionality~ into the target type? The only operation that should do it, that I know of, involved homomorphic mapped types - and in this case, indexed access types are not involved directly.